prov/sm2/src/sm2_progress.c

```diff
@@ -202,6 +203,278 @@ static int sm2_progress_recv_msg(struct sm2_ep *ep,
+		sm2_freestack_push(ep, xfer_entry);
+		return FI_SUCCESS;
+	}
+
+	ret = sm2_rma_cmd_fill_sar_xfer(xfer_entry, ctx);
+	if (!ret)
+		sm2_fifo_write(ep, ctx->peer_gid, xfer_entry);
+	else
+		sm2_rma_handle_local_error(ep, xfer_entry, ctx, ret);
+
+	return FI_SUCCESS;
+}
+
+static int sm2_progress_rma_write_origin(struct sm2_ep *ep,
+					 struct sm2_xfer_entry *xfer_entry)
+{
+	struct sm2_cmd_sar_rma_msg *cmd_rma = (void *) xfer_entry->user_data;
+	struct sm2_sar_ctx *ctx = (void *) cmd_rma->sar_hdr.context;
+	bool tx_complete, rx_complete;
+	int ret;
+
+	if (!ctx) {
+		/* inject message being returned. */
+		sm2_freestack_push(ep, xfer_entry);
+		return FI_SUCCESS;
+	}
+	ctx->msgs_in_flight--;
+
+	ctx->status_flags |=
+		cmd_rma->sar_hdr.proto_flags & FI_SM2_SAR_ERROR_FLAG;
+	if (ctx->status_flags & FI_SM2_SAR_ERROR_FLAG) {
+		sm2_rma_handle_remote_error(ep, xfer_entry, ctx);
+		return FI_SUCCESS;
+	}
+
+	tx_complete = ctx->bytes_sent == ctx->bytes_total;
+	rx_complete =
+		FI_SM2_SAR_LAST_MESSAGE_FLAG & cmd_rma->sar_hdr.proto_flags;
+	if (rx_complete) {
+		if (0 == (ctx->status_flags & FI_SM2_SAR_STATUS_COMPLETED)) {
+			ret = sm2_complete_tx(
+				ep, (void *) xfer_entry->hdr.context,
+				xfer_entry->hdr.op, xfer_entry->hdr.op_flags);
+			if (ret)
+				FI_WARN(&sm2_prov, FI_EP_RDM,
+					"Error generating completion for RMA "
+					"write.");
+			else
+				ctx->status_flags |=
+					FI_SM2_SAR_STATUS_COMPLETED;
+		}
+		sm2_free_rma_ctx(ctx);
+		sm2_freestack_push(ep, xfer_entry);
+	} else if (tx_complete) {
+		sm2_freestack_push(ep, xfer_entry);
+	} else {
+		ret = sm2_rma_cmd_fill_sar_xfer(xfer_entry, ctx);
+		if (ret)
+			sm2_rma_handle_local_error(ep, xfer_entry, ctx, ret);
+		else
+			sm2_fifo_write(ep, ctx->peer_gid, xfer_entry);
+	}
+
+	return FI_SUCCESS;
+}
+static int sm2_progress_rma_write_remote(struct sm2_ep *ep,
+					 struct sm2_xfer_entry *xfer_entry)
+{
+	int ret = 0, jv;
+	struct sm2_cmd_sar_rma_msg *cmd_rma;
+	struct fi_rma_iov *iov;
+
+	uint8_t *src;
+	void *buf_for_completion = NULL;
+	ssize_t total_bytes_for_completion = 0;
+	uintptr_t hmem_addr;
+	struct sm2_domain *domain;
+	struct ofi_mr *mr;
+
+	domain = container_of(ep->util_ep.domain, struct sm2_domain,
+			      util_domain);
+
+	cmd_rma = (struct sm2_cmd_sar_rma_msg *) xfer_entry->user_data;
+
+	/* copy data from packet to destination */
+	src = (uint8_t *) &cmd_rma->user_data;
+	iov = &cmd_rma->rma_iov[0];
+
+	for (jv = 0; jv < SM2_IOV_LIMIT && iov->len > 0; jv++, iov++) {
+		hmem_addr = (uintptr_t) iov->addr;
+		ret = ofi_mr_map_verify(&domain->util_domain.mr_map, &hmem_addr,
+					iov->len, iov->key, FI_REMOTE_WRITE,
+					(void **) &mr);
+		if (ret) {
+			cmd_rma->sar_hdr.proto_flags |= FI_SM2_SAR_ERROR_FLAG;
+			break;
+		}
+		ret = ofi_copy_to_hmem(mr->iface, mr->device,
+				       (void *) hmem_addr, src, iov->len);
+		src += iov->len;
+		assert((long) (src - (uint8_t *) &cmd_rma->user_data) <=
+		       SM2_RMA_INJECT_SIZE);
+		if (ret) {
```

Comment:
Addressed.  Sequence for this case is:
1. Remote (the ep in this progress function) emits error to CQ
2. Origin emits error to CQ
3. Origin cleans up context.

As long as remote has no context and only 1 message is in flight at a time, this is sufficient.

If we want more than 1 message in flight at a time, we need some state at remote to prevent a subsequent message from  succeeding and accidentally submitting 2 CQs.